### PR TITLE
Refactor: Burn Import use BurnGraph

### LINKED ISSUE #336: ONNX: Develop Intermediate Representation for Burn Code Generation
### Feature Description

Design and implement an intermediate representation (IR) specifically for Burn code generation in the ONNX framework.

### Motivation for Feature

Currently, the burn-import process converts ONNX protobufs to an intermediate representation (IR) closely resembling the ONNX data. This IR is subsequently converted to Burn source code. To streamline this process, it would be beneficial to introduce an additional IR layer, as follows: ONNX -> ONNX IR -> Burn IR -> Burn Source Code.

The data structure might look like this:

```rust
enum BurnNode {
    MatMul(MatMulArgs),
    ...
}

struct MatMulArgs {
    lhs: TensorNode,
    rhs: TensorNode,
}
```

### PR DESCRIPTION
**Pull request by nathanielsimard**
Use an internal representation for burn modules and tensor operations.

fix #336

### CODE REVIEW

File: burn-import/src/onnx/to_burn.rs [hunk 1]
```diff
@@ -96,558 +95,213 @@ impl ModelGen {
...
-
-        }
+        graph
     }
 
-    /// Model implementation code
-    fn implementation(&self, imports: &mut HashSet<String>) -> TokenStream {
-        let forward_method = self.forward_method(imports);
+    fn matmul_conversion(node: Node) -> MatmulNode {
+        println!("Matmul Conversion");
```
Comment (antimora) [inline burn-import/src/onnx/to_burn.rs:159]: The same comment about print statement.

File: burn-import/src/onnx/to_burn.rs [hunk 2]
```diff
@@ -96,558 +95,213 @@ impl ModelGen {
...
-            NodeType::LogSoftmax => {
-                imports.insert("burn::tensor::activation::log_softmax".to_string());
-                let dim = log_softmax_config(node);
-
-                quote! { log_softmax(#input1, #dim) }
-            }
-            NodeType::Flatten => {
-                let (start_dim, end_dim) = flatten_config(node);
+    fn linear_conversion<PS: PrecisionSettings>(mut node: Node) -> LinearNode<PS> {
+        println!("Linear Conversion");
```
Comment (antimora) [inline burn-import/src/onnx/to_burn.rs:191]: The same comment about print statement.

File: burn-import/src/onnx/to_burn.rs [hunk 3]
```diff
@@ -96,558 +95,213 @@ impl ModelGen {
...
-                 )*)
-            }
-        }
-    }
-
-    /// Generates source code for the initialization method
-    fn new_method(&self) -> TokenStream {
-        let initialization_fields = self.initialization_fields();
+    fn conv2d_conversion<PS: PrecisionSettings>(mut node: Node) -> Conv2dNode<PS> {
+        println!("Conv2d conversion");
```
Comment (antimora) [inline burn-import/src/onnx/to_burn.rs:250]: The same comment about print statement.

File: burn-import/tests/data/model1/model1.rs [hunk 4]
```diff
@@ -1,67 +1,59 @@
...
+use burn::nn::LinearConfig;
 use burn::{
     module::Module,
-    nn,
-    record::{DefaultRecorder, Recorder},
     tensor::{backend::Backend, Tensor},
 };
 
-pub const INPUT1_SHAPE: [usize; 4usize] = [1usize, 1usize, 8usize, 8usize];
-pub const OUTPUT1_SHAPE: [usize; 2usize] = [1usize, 288usize];
```
Comment (antimora) [inline burn-import/tests/data/model1/model1.rs:14]: Do we still want to provide some facility to know the input shape? The user may have to guess or know in advance but it's part of the model.

**Review by antimora — APPROVED** [review]
Looks great! As I was reviewing I became excited about your simplification. I'd be so much easier now to add more nodes. I guess your work tackles both #334 and #336.

I approved it but please see my comments with questions.

**Review by antimora — APPROVED** [review]
Looks great! As I was reviewing I became excited about your simplification. I'd be so much easier now to add more nodes. I guess your work tackles both #334 and #336.

I approved it but please see my comments with questions.

File: burn-import/tests/data/model1/model1.rs (hunk 4 above)
Comment (nathanielsimard) [reply burn-import/tests/data/model1/model1.rs:14]: The input and output shapes may vary, if you want to use a batch size of 1 or 32 it shouldn't matter, so I prefer not exporting it, making its usage similar to other burn modules. The best solution would be to export comments explaining each dimension (like batch_size, seq_length, channel, width, etc.) but I'm not sure this information is present in ONNX.

Comment (antimora) [reply burn-import/tests/data/model1/model1.rs:14]: We could just document it in the source somewhere. I am not sure really. Or maybe we could provide some tooling that summarizes the model like this: http://www.xavierdupre.fr/app/onnxruntime/helpsphinx/auto_examples/plot_load_and_predict.html

We could think of a solution and do it later. I am okay with this.

**Review by antimora — APPROVED** [review]

**Pull request merged**